tests/integration/test_api.py

```diff
@@ -0,0 +1,336 @@
+        "count": 1,
+        "questions": [
+            {
+                "barcode": "1",
+                "type": "add-binary",
+                "value": "Seeds",
+                "question": "Does the product belong to this category?",
+                "insight_id": insight_id,
+                "insight_type": "category",
+            }
+        ],
+        "status": "found",
+    }
+
+
+def test_barcode_question():
+    cl = client()
+    result = cl.simulate_get("/api/v1/questions/2")
+
+    assert result.status_code == 200
+    assert result.json == {"questions": [], "status": "no_questions"}
+
+    result = cl.simulate_get("/api/v1/questions/1")
+
+    assert result.status_code == 200
+    assert result.json == {
+        "questions": [
+            {
+                "barcode": "1",
+                "type": "add-binary",
+                "value": "Seeds",
+                "question": "Does the product belong to this category?",
+                "insight_id": insight_id,
+                "insight_type": "category",
+            }
+        ],
+        "status": "found",
+    }
+
+
+def test_annotate_insight_authenticated():
+    cl = client()
+    result = cl.simulate_post(
+        "/api/v1/insights/annotate",
+        params={
+            "insight_id": insight_id,
+            "annotation": -1,
+        },
+        headers={"Authorization": "Basic " + base64.b64encode(b"a:b").decode("ascii")},
+    )
+
+    assert result.status_code == 200
+    assert result.json == {"description": "the annotation was saved", "status": "saved"}
+
+    # For authenticated users we expect the insight to be validated directly, tracking the username of the annotator.
+    votes = list(AnnotationVote.select())
+    assert len(votes) == 0
+
+    insight = next(
+        ProductInsight.select()
+        .where(ProductInsight.id == insight_id)
+        .dicts()
+        .iterator()
+    )
+    assert insight.items() > {"username": "a", "annotation": -1, "n_votes": 0}.items()
+    assert "completed_at" in insight
+
+
+def test_annotate_insight_not_enough_votes():
+    cl = client()
+    result = cl.simulate_post(
+        "/api/v1/insights/annotate",
+        params={
+            "insight_id": insight_id,
+            "annotation": -1,
+            "device_id": "voter1",
+        },
+    )
+
+    assert result.status_code == 200
+    assert result.json == {
+        "description": "the annotation vote was saved",
+        "status": "vote_saved",
+    }
+
+    # For non-authenticated users we expect the insight to not be validated, with only a vote being cast.
+    votes = list(AnnotationVote.select().dicts())
+    assert len(votes) == 1
+
+    assert votes[0]["value"] == -1
+    assert votes[0]["username"] is None
+    assert votes[0]["device_id"] == "voter1"
+
+    insight = next(
+        ProductInsight.select()
+        .where(ProductInsight.id == insight_id)
+        .dicts()
+        .iterator()
+    )
+
+    assert all(
+        [
```

Comment:
you don't need to put list "[" in all, you can directly write:
`all(val is None for key, val in insight.items() if key in {"username", "completed_at", "annotation"})`

I would even have written:
`not any(insight[key] for key in ("username", "completed_at", "annotation"))`

Reply:
Done, thanks!